c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,338 @@
+  gs.add<&Ftrl::get_n_epochs, &Ftrl::set_n_epochs>("n_epochs", "Number of epochs to train for.\n");
+  gs.add<&Ftrl::get_inter, &Ftrl::set_inter>("inter", "If feature interactions to be used or not.\n");
+  gs.add<&Ftrl::get_hash_type, &Ftrl::set_hash_type>("hash_type",
+    "Hashing method to use for strings.\n"
+    "`0` - std::hash;\n"
+    "`1` - Murmur2;\n"
+    "`2` - Murmur3.\n");
+  gs.add<&Ftrl::get_seed, &Ftrl::set_seed>("seed", "Seed to be used for Murmur hash functions.\n");
+
+  mm.add<&Ftrl::fit, args_fit>();
+  mm.add<&Ftrl::predict, args_predict>();
+}
+
+
+PKArgs Ftrl::Type::args_fit(1, 0, 0, false, false, {"df_train"}, "fit",
+R"(fit(self, frame)
+--
+
+Train an FTRL model on a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame to be trained on, last column is treated as `target`.
+
+Returns
+----------
+    A new Frame of shape `(d, 2)` containing `z` and `n` model coefficients.
+)");
+
+
+void Ftrl::fit(const PKArgs& args) {
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+  DataTable* dt_train = args[0].to_frame();
+  fm->fit(dt_train);
+}
+
+
+PKArgs Ftrl::Type::args_predict(1, 0, 0, false, false, {"df_test"}, "predict",
+R"(predict(self, frame)
+--
+
+Make predictions for a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame of shape `(nrows, ncols)` to make predictions for. It must have one 
+    column less than the training dataset.
+
+Returns
+----------
+    A new `Frame` of shape `(nrows, 1)` with a prediction for each row.
+)");
+
+
+oobj Ftrl::predict(const PKArgs& args) {
+  if (fm->get_model() == nullptr) {
+    throw ValueError() << "There is no trained model in place.\n"
+                       << "To make predictions, the model must be either trained first, or "
+                       << "set.";
+  }
+
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+
+  DataTable* dt_test = args[0].to_frame();
+  DataTable* dt_target = fm->predict(dt_test).release();
+  py::Frame* df_target = py::Frame::from_datatable(dt_target);
+
+  return df_target;
+}
+
+
+void Ftrl::set_model(robj model) {
+  if (!model.is_frame()) {
+    throw TypeError() << "`model` must be a frame, not "
+        << model.typeobj();
+  }
+
+  DataTable* dt_model_in = model.to_frame();
+  const std::vector<std::string> model_cols_in = dt_model_in->get_names();
+  if (dt_model_in->nrows == fm->get_d() && dt_model_in->ncols == 2 &&
+    dt_model_in->columns[0]->stype() == SType::FLOAT64 &&
+    dt_model_in->columns[1]->stype() == SType::FLOAT64 &&
+    model_cols_in == FtrlModel::model_cols) {
+    fm->set_model(dt_model_in);
+  } else {
+    throw ValueError() << "FTRL model frame must have " << fm->get_d() <<" rows, and" <<
+                          "2 columns named `z` and `n`, " <<
+                          "both columns must be of `FLOAT64` type.";
+  }
```

Comment:
it would be better to split this into multiple checks, each with a more specific error message. For example, `"FTRL model frame must have " << fm->get_d() << " rows and 2 columns, whereas your frame has shape " << nrows << " rows and " << ncols << " columns"`. And so on.
This is of course more to write, but it will help the user to diagnose their problems.